File: backend/apps/projects/schemas.py
from datetime import date, datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from ninja import Schema
from pydantic import Field


class UserSummarySchema(Schema):
    id: int
    email: str
    first_name: str
    last_name: str


class ProjectTeamMemberCreateSchema(Schema):
    user_id: int
    role: str = Field(default="", max_length=150)
    allocation_percentage: int = Field(default=100, ge=1, le=100)


class ProjectTeamMemberSchema(Schema):
    id: UUID
    user: UserSummarySchema
    role: str
    allocation_percentage: int
    is_active: bool
    created_at: datetime


class ProjectMilestoneCreateSchema(Schema):
    title: str = Field(min_length=1, max_length=250)
    description: str = ""
    status: str = "pending"
    start_date: date | None = None
    due_date: date | None = None
    progress: int = Field(default=0, ge=0, le=100)
    sort_order: int = Field(default=0, ge=0)
    amount: Decimal = Field(default=Decimal("0.00"), ge=0)


class ProjectMilestoneUpdateSchema(ProjectMilestoneCreateSchema):
    pass


class ProjectMilestoneSchema(Schema):
    id: UUID
    title: str
    description: str
    status: str
    start_date: date | None
    due_date: date | None
    completed_at: datetime | None
    progress: int
    sort_order: int
    amount: Decimal
    created_at: datetime
    updated_at: datetime


class ProjectNoteCreateSchema(Schema):
    content: str = Field(min_length=1)
    is_pinned: bool = False
    is_client_visible: bool = False


class ProjectNoteSchema(Schema):
    id: UUID
    content: str
    is_pinned: bool
    is_client_visible: bool
    created_at: datetime


class ProjectEventSchema(Schema):
    id: UUID
    event_type: str
    description: str
    metadata: dict[str, Any]
    created_at: datetime


class ProjectCreateSchema(Schema):
    client_id: UUID
    project_manager_id: int | None = None
    title: str = Field(min_length=1, max_length=250)
    description: str = ""
    status: str = "planning"
    priority: str = "normal"
    budget: Decimal = Field(default=Decimal("0.00"), ge=0)
    currency: str = Field(default="LKR", min_length=3, max_length=3)
    start_date: date | None = None
    deadline: date | None = None
    progress: int = Field(default=0, ge=0, le=100)
    repository_url: str = ""
    staging_url: str = ""
    production_url: str = ""
    notes: str = ""
    tags: list[str] = Field(default_factory=list)


class ProjectUpdateSchema(ProjectCreateSchema):
    client_id: UUID | None = None


class ProjectSchema(Schema):
    id: UUID
    project_code: str
    client_id: UUID
    client_name: str
    quotation_id: UUID | None
    title: str
    description: str
    status: str
    priority: str
    budget: Decimal
    currency: str
    start_date: date | None
    deadline: date | None
    completed_at: datetime | None
    progress: int
    project_manager: UserSummarySchema | None
    repository_url: str
    staging_url: str
    production_url: str
    notes: str
    tags: list[str]
    team_members: list[ProjectTeamMemberSchema]
    milestones: list[ProjectMilestoneSchema]
    project_notes: list[ProjectNoteSchema]
    events: list[ProjectEventSchema]
    created_at: datetime
    updated_at: datetime

File: apps/dashboard/src/features/projects/schemas.ts
import {
  z,
} from "zod";

import {
  milestoneStatuses,
  projectPriorities,
  projectStatuses,
} from "./types";

const userSummarySchema = z.object({
  id: z.number().int(),
  email: z.string(),
  first_name: z.string(),
  last_name: z.string(),
});

const teamMemberSchema = z.object({
  id: z.string().uuid(),
  user: userSummarySchema,
  role: z.string(),
  allocation_percentage:
    z.number().int().min(1).max(100),
  is_active: z.boolean(),
  created_at: z.string(),
});

const milestoneSchema = z.object({
  id: z.string().uuid(),
  title: z.string(),
  description: z.string(),
  status: z.enum(milestoneStatuses),
  start_date: z.string().nullable(),
  due_date: z.string().nullable(),
  completed_at: z.string().nullable(),
  progress: z.number().int().min(0).max(100),
  sort_order: z.number().int().nonnegative(),
  amount: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
});

const projectNoteSchema = z.object({
  id: z.string().uuid(),
  content: z.string(),
  is_pinned: z.boolean(),
  is_client_visible: z.boolean(),
  created_at: z.string(),
});

const projectEventSchema = z.object({
  id: z.string().uuid(),
  event_type: z.string(),
  description: z.string(),
  metadata: z.record(
    z.string(),
    z.unknown(),
  ),
  created_at: z.string(),
});

export const projectSchema = z.object({
  id: z.string().uuid(),
  project_code: z.string(),
  client_id: z.string().uuid(),
  client_name: z.string(),
  quotation_id: z.string().uuid().nullable(),
  title: z.string(),
  description: z.string(),
  status: z.enum(projectStatuses),
  priority: z.enum(projectPriorities),
  budget: z.string(),
  currency: z.string(),
  start_date: z.string().nullable(),
  deadline: z.string().nullable(),
  completed_at: z.string().nullable(),
  progress: z.number().int().min(0).max(100),
  project_manager: userSummarySchema.nullable(),
  repository_url: z.string(),
  staging_url: z.string(),
  production_url: z.string(),
  notes: z.string(),
  tags: z.array(z.string()),
  team_members: z.array(teamMemberSchema),
  milestones: z.array(milestoneSchema),
  project_notes: z.array(projectNoteSchema),
  events: z.array(projectEventSchema),
  created_at: z.string(),
  updated_at: z.string(),
});

export const paginatedProjectsSchema = z.object({
  items: z.array(projectSchema),
  pagination: z.object({
    page: z.number().int().positive(),
    page_size: z.number().int().positive(),
    total_items: z.number().int().nonnegative(),
    total_pages: z.number().int().nonnegative(),
  }),
});
